datafusion/core/src/config.rs

```diff
@@ -17,616 +17,517 @@
-pub const OPT_ENABLE_ROUND_ROBIN_REPARTITION: &str =
-    "datafusion.optimizer.enable_round_robin_repartition";
+        /// Type of `TableProvider` to use when loading `default` schema
+        pub format: Option<String>, default = None
 
-/// Definition of a configuration option
-pub struct ConfigDefinition {
-    /// key used to identifier this configuration option
-    key: String,
-    /// Description to be used in generated documentation
-    description: String,
-    /// Data type of this option
-    data_type: DataType,
-    /// Default value
-    default_value: ScalarValue,
+        /// If the file has a header
+        pub has_header: bool, default = false
+    }
 }
 
-macro_rules! get_conf_value {
-    ($SELF: expr, $TPE: ident, $KEY: expr, $TPE_NAME: expr) => {
-        match $SELF.get($KEY) {
-            Some(ScalarValue::$TPE(v)) => v,
-            Some(v) => {
-                warn!(
-                    "Config type mismatch for {}. Expected: {}, got: {:?}",
-                    $KEY, $TPE_NAME, &v
-                );
-                None
-            }
-            None => None,
-        }
-    };
+config_namespace! {
+    pub struct ExecutionOptions {
+        /// Default batch size while creating new batches, it's especially useful for
+        /// buffer-in-memory batches since creating tiny batches would results in too much
+        /// metadata memory consumption
+        pub batch_size: usize, default = 8192
+
+        /// When set to true, record batches will be examined between each operator and
+        /// small batches will be coalesced into larger batches. This is helpful when there
+        /// are highly selective filters or joins that could produce tiny output batches. The
+        /// target batch size is determined by the configuration setting
+        pub coalesce_batches: bool, default = true
+
+        /// Target batch size when coalescing batches. Used in conjunction with `coalesce_batches`
+        pub coalesce_target_batch_size: usize, default = 4096
+
+        /// Should DataFusion collect statistics after listing files
+        pub collect_statistics: bool, default = false
+
+        /// Number of partitions for query execution. Increasing partitions can increase
+        /// concurrency. Defaults to the number of cpu cores on the system
+        pub target_partitions: usize, default = num_cpus::get()
+
+        /// The default time zone
+        ///
+        /// Some functions, e.g. EXTRACT(HOUR from SOME_TIME), shift the underlying datetime
+        /// according to this time zone, and then extract the hour
+        pub time_zone: Option<String>, default = Some("+00:00".into())
+
+        /// When set to true, the physical plan optimizer will try to add round robin
+        /// repartition to increase parallelism to leverage more CPU cores
+        pub round_robin_repartition: bool, default = true
+
+        /// Parquet options
+        pub parquet: ParquetOptions, default = Default::default()
```

Comment:
This nesting is a source of non-trivial additional complexity, but I made it work so at least we have support for this in the future.